BindingInfo spans point to the derive rather than the individual field

The spans of `__binding_0` etc. and of `field.span()` point to the derive, which makes it hard to see which field is causing an error.

Would be good to use `located_at` to make them point to the relevant field.

Example proc macro:
```rust
use proc_macro2::TokenStream;
use quote::{quote, quote_spanned};
use syn::spanned::Spanned;
use synstructure::Structure;

synstructure::decl_derive!([Print] => derive_print);

fn derive_print(input: Structure<'_>) -> TokenStream { 
    let mut body = Vec::new();
    let span = input.ast().span();
    for variant in input.variants() {
        let fields: Vec<_> = variant
            .bindings()
            .iter()
            .map(|f| quote_spanned! { f.span() => { eprintln!("{}", #f) }})
            .collect();
        let pat = variant.pat();
        body.push(quote_spanned! {span=>
            #pat => { #(#fields);* }
        });
    }

    input.gen_impl(quote_spanned! {span=>
        gen impl Print for @Self {
            fn print(&self) {
                match *self {
                    #(#body)*
                }
            }
        }
    })
}

#[proc_macro]
pub fn my_proc_impl(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let input: proc_macro2::TokenStream = input.into();
    quote!(println!("Answer: {:?}", #input)).into()
}
```

And user code:
```
use pm_exp::Print;

trait Print {
    fn print(&self);
}

#[derive(Print)]
struct Xyz {
    x: u32,
    y: *const (),
}
```

The error is
```
error[E0277]: `*const ()` doesn't implement `std::fmt::Display`
 --> tests/foo.rs:7:10
  |
7 | #[derive(Print)]
  |          ^^^^^ `*const ()` cannot be formatted with the default formatter
  |
  = help: the trait `std::fmt::Display` is not implemented for `*const ()`
  = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
  = note: this error originates in the macro `$crate::format_args_nl` which comes from the expansion of the derive macro `Print` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `pm-exp` (test "foo") due to 1 previous error
```

Would have been better if the squigglies point to the field `y` rather than to the struct.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -30,7 +30,7 @@ jobs:
             target: wasm32-unknown-unknown
           - rust: nightly
             name: WASI
-            target: wasm32-wasi
+            target: wasm32-wasip1
           - rust: stable
             name: Windows
             os: windows

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -164,6 +164,7 @@ extern crate proc_macro;
 use std::collections::HashSet;
 
 use syn::parse::{ParseStream, Parser};
+use syn::spanned::Spanned;
 use syn::visit::{self, Visit};
 use syn::{
     braced, punctuated, token, Attribute, Data, DeriveInput, Error, Expr, Field, Fields,
@@ -488,10 +489,12 @@ impl<'a> VariantInfo<'a> {
                     .into_iter()
                     .enumerate()
                     .map(|(i, field)| {
+                        // XXX: binding_span has to be call_site to avoid privacy
+                        // when deriving on private fields, but be located at the field
+                        // span for nicer diagnostics.
+                        let binding_span = Span::call_site().located_at(field.span());
                         BindingInfo {
-                            // XXX: This has to be call_site to avoid privacy
-                            // when deriving on private fields.
-                            binding: format_ident!("__binding_{}", i),
+                            binding: format_ident!("__binding_{}", i, span = binding_span),
                             style: BindStyle::Ref,
                             field,
                             generics,

--- a/src/macros.rs
+++ b/src/macros.rs
@@ -4,8 +4,8 @@
 
 // Re-exports used by the decl_derive! and test_derive!
 pub use proc_macro2::TokenStream as TokenStream2;
-pub use syn::{parse2, parse_str, DeriveInput};
 pub use quote::quote;
+pub use syn::{parse2, parse_str, DeriveInput};
 
 #[cfg(all(
     not(all(target_arch = "wasm32", any(target_os = "unknown", target_os = "wasi"))),

